### PR TITLE
fix various const eval errors

### LINKED ISSUE #33275: `i8::MIN as f32`: warning: attempted to negate with overflow, #[warn(const_err)] on by default
### STR

``` rust
// cast.rs
use std::i8;

fn main() {
    i8::MIN as f32;
}
```

```
$ rustc cast.rs
cast.rs:5:5: 5:19 warning: attempted to negate with overflow, #[warn(const_err)] on by default
cast.rs:5     i8::MIN as f32;
              ^~~~~~~~~~~~~~
```

The "evaluated" expression `-128 as f32` doesn't trigger the warning. But if you expand the definition of `i8::MIN` i.e. `((-1 as i8) << 8 - 1) as f32` you also get a warning.

Neither of these three equivalent expressions should trigger a warning.

### Metadata

```
$ rustc -V
rustc 1.10.0-nightly (cda7c1cf2 2016-04-27)
```

cc @oli-obk might be related to #33214

### LINKED ISSUE #33291: Warning on u8 -> char casts; warning: can't cast this type, #[warn(const_err)]
This is a new warning in recent nightlies, for u8 -> char casts.

```rust
src/bin.rs:71:13: 71:25 warning: can't cast this type, #[warn(const_err)] on by default
src/bin.rs:71     if c >= 33u8 as char && c < 128u8 as char {
                          ^~~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by oli-obk**
These were found after const_evaluating arbitrary expressions and linting if the const evaluator failed

fixes #33275 (int -> float casts for negative ints)
fixes #33291 (int -> char casts (new! wasn't allowed in constants until this PR))

r? @eddyb 

cc @bluss @japaric

### CODE REVIEW

**Comment by oli-obk** [comment]
hmm... where's bors? or is @japaric not a reviewer?

r? @eddyb

File: src/librustc_const_eval/eval.rs [hunk 1]
```diff
@@ -1089,23 +1077,19 @@ fn cast_const_int<'tcx>(tcx: &TyCtxt<'tcx>, val: ConstInt, ty: ty::Ty) -> CastRe
...
         },
-        ty::TyFloat(ast::FloatTy::F64) => Ok(Float(val.to_u64().unwrap() as f64)),
-        ty::TyFloat(ast::FloatTy::F32) if val.is_negative() => {
-            let val = (-val).map_err(Math)?;
-            let val = val.to_u64().unwrap() as f32;
-            let val = -val;
-            Ok(Float(val as f64))
+        ty::TyFloat(ast::FloatTy::F32) => match val.erase_type() {
+            Infer(u) => Ok(Float(u as f32 as f64)),
+            InferSigned(i) => Ok(Float(i as f32 as f64)),
```
Comment (eddyb) [inline src/librustc_const_eval/eval.rs]: This reminds me, we need 2 variants, one for `f32` and one for `f64`.

File: src/librustc_const_eval/eval.rs [hunk 2]
```diff
@@ -377,13 +377,6 @@ pub enum ErrKind {
     NotOn(ConstVal),
     CallOn(ConstVal),
 
-    NegateWithOverflow(i64),
```
Comment (japaric) [inline src/librustc_const_eval/eval.rs]: These are gone because they were unused?

File: src/librustc_const_eval/eval.rs [hunk 3]
```diff
@@ -482,6 +469,7 @@ impl ConstEvalErr {
             },
             BadType(ref i) => format!("value of wrong type: {:?}", i).into_cow(),
             ErroneousReferencedConstant(_) => "could not evaluate referenced constant".into_cow(),
+            BadCharValue => "invalid numeric value for char".into_cow(),
```
Comment (japaric) [inline src/librustc_const_eval/eval.rs]: Could you add tests for this warnings? Probably `some_i8 as char` and `u32_thats_not_a_valid_char as char`

File: src/test/run-pass/const-err.rs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
```
Comment (japaric) [inline src/test/run-pass/const-err.rs]: nit: year

File: src/librustc_const_eval/eval.rs [hunk 5]
```diff
@@ -1089,23 +1077,19 @@ fn cast_const_int<'tcx>(tcx: &TyCtxt<'tcx>, val: ConstInt, ty: ty::Ty) -> CastRe
...
-            // FIXME: this could probably be prettier
-            // there's no easy way to turn an `Infer` into a f64
-            let val = (-val).map_err(Math)?;
-            let val = val.to_u64().unwrap() as f64;
-            let val = -val;
-            Ok(Float(val))
+        ty::TyFloat(ast::FloatTy::F64) => match val.erase_type() {
+            Infer(u) => Ok(Float(u as f64)),
+            InferSigned(i) => Ok(Float(i as f64)),
+            _ => unreachable!(),
```
Comment (japaric) [inline src/librustc_const_eval/eval.rs]: nit: I think there's a `bug!` macro for these cases?

Comment (eddyb) [inline src/librustc_const_eval/eval.rs]: Or `span_bug!` if you happen to have a `Span` around.

**Comment by japaric** [comment]
> hmm... where's bors? or is @japaric not a reviewer?

I have r+ rights, but I think I'm not be registered with the highfive bot.

> fixes #33291 (int -> float casts for negative ints)
> fixes #33275 (int -> char casts (new! wasn't allowed in constants until this PR))

the descriptions seem to be backward.

@oli-obk r=me with added tests cases :smile:.

File: src/librustc_const_eval/eval.rs (hunk 2 above)
Comment (oli-obk) [inline src/librustc_const_eval/eval.rs]: yea, just a drive-by-cleanup

**Comment by oli-obk** [comment]
> the descriptions seem to be backward.

whoops :) fixed

> added tests cases

hmm... when I tried to create a cfail test for the error message I found out that it's the wrong message and the wrong logic. It's not possible to cast anything but a `u8` to a `char`. And the regular rustc error message even says so. So I made sure that the value is either a `u8` or at least can be inferred into a `u8`. But that caused some fallout in the `infer` function, because the cast-function doesn't have a span, but the `infer` function expects a span to report any errors. I changed this to only return an `ErrKind` and to attach the span in `eval_const_expr_partial`.

**Comment by japaric** [comment]
> It's not possible to cast anything but a u8 to a char. And the regular rustc error message even says so.

Oh, right. I didn't remember that was the case.

> Note that typeck catches this issue, so we can only test for the const_eval error message by evaluating a pattern or array length

So typeck doesn't reject this cast when it's in an array length. That's ... interesting.

File: src/test/compile-fail/const-eval-overflow-4b.rs [hunk 6]
```diff
@@ -26,10 +25,15 @@ const A_I8_T
...
 
-fn main() {
-    foo(&A_I8_T[..]);
-}
 
-fn foo<T:fmt::Debug>(x: T) {
-    println!("{:?}", x);
-}
+const A_CHAR_USIZE
+    : [u32; 5u8 as char as usize]
```
Comment (japaric) [inline src/test/compile-fail/const-eval-overflow-4b.rs]: Yay, more const casting possibilities.

**Comment by japaric** [comment]
@bors r+

Thanks @oli-obk!

**Comment by durka** [comment]
another one: `b"x" as &[u8]`

**Comment by oli-obk** [comment]
@durka I can fix this easily. This is a mis-use of `as`, so we should probably start warning on it anyway

**Comment by durka** [comment]
How is it a misuse? It's unsizing the `&[u8; 1]` to a `&[u8]`. But more than that, the message is bad -- it says "cannot cast this type" while the code compiles, so it certainly does cast the type, and it doesn't give the user any information about why the cast might be invalid.

**Comment by oli-obk** [comment]
shouldn't a simple `as_ref` suffice?

```rust
#![feature(type_ascription)]

fn main() {
    let x = b"foo".as_ref();
    let y = x : &[u8];
}
```

I was under the impression that `as` should only be used for casting when there's no other way.

**Comment by eddyb** [comment]
@oli-obk Calling `as_ref` in non-generic code is an anti-pattern. Also, it doesn't matter if `as` is bad style, sometimes it's the simplest option (given that ascription doesn't coerce).
`&b"foo"[..]` works but it's a bit uglier.
The problem is that the warning is wrong: the coercion/cast works fine, even if it's not implemented in `const_eval`.

**Pull request merged**